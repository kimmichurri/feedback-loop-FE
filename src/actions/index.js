export const hasError = (message) => ({
  type: 'HAS_ERROR',
  message
})

export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
})

export const setSurvey = (survey) => ({
  type: 'SET_SURVEY',
  survey
})

export const setCohorts = (cohorts) => ({
  type: 'SET_COHORTS',
  cohorts
})

export const setUser = (user) => ({
  type: 'SET_USER',
  user
})

export const setCurrentCohort = (cohort) => ({
  type: 'SET_CURRENT_COHORT',
  cohort
})