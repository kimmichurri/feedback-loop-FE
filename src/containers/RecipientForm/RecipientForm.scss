.recipients-form-wrapper {
  display: flex;
  flex-direction: column;
}

.students-display {
  display: flex;
  flex-wrap: wrap;
  background-color: $accent-red;
}

.groups-wrapper {
  background-color: $main-blue;
  height: 200px;
}

.student-nametag {
  background-color: $light-grey;
  border-radius: 4px;
  height: 75px;
  margin: 1rem;
  width: 100px;
  &:hover {
    color: $main-grey;
    cursor: grab;
    font-weight: bolder;
    background-color: $light-grey-hover;
    box-shadow: 1px 1px 2px $main-grey;
  }
}

.student-name {
  padding: 4px;
}

.recipients-form-title,
.student-selector,
.student-name,
.drop-down {
  font-family: $main-font
}

.drop-down {
  height: 40px;
  margin: auto;
  width: 200px;
  &:focus {
    outline: 2px solid $main-blue;
  }
}

.recipients-button {
  background-color: $main-blue;
  border-radius: 50px;
  border: none;
  color: $accent-teal;
  font-family: $main-font;
  font-weight: bold;
  font-size: 16px;
  height: 50px;
  margin: 20px auto;
  width: 200px;
  &:hover {
    cursor: pointer;
    background-color: $main-blue-hover;
  }
}