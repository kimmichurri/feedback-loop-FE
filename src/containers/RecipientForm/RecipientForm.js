import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { handlePost } from '../../thunks/handlePost'
import { setCurrentCohort } from '../../actions'
import cogoToast from 'cogo-toast';

export class RecipientForm extends Component {
  constructor() {
    super()
    this.state = {
      cohort_id: 0,
      program: 'both',
      draggedStudent: {},
      group: [],
      displayTeams: "none"
    }
  }

  handleProgram = (e) => {
    this.setState({
      program: e.target.value
    })
  }

  handleCohort = (e) => {
    this.setState({
      cohort_id: e.target.value
    })
  }

  handleAssignGroups = async () => {
    const { cohort_id, program } = this.state
    let url
    this.state.program === "both"
    ? url = `https://turing-feedback-api.herokuapp.com/api/v1/students?cohort=${cohort_id}`
    : url = `https://turing-feedback-api.herokuapp.com/api/v1/students?cohort=${cohort_id}&&program=${program}`
    const response = await fetch(url)
    const cohort = await response.json()
    await this.props.setCurrentCohort(cohort)
    this.setState({
      displayTeams: "flex"
    })
  }

  postSurvey = () => {
    if(this.state.group.length < 2) {
      this.sendToast('There must be at least two students in a group')
    } else {
      const membersIds = []
      this.state.group.forEach(student => {
        membersIds.push(student.id)
      })
      const { cohort_id } = this.state
      const { survey } = this.props
      const url = "https://turing-feedback-api.herokuapp.com/api/v1/surveys"
      const options = {
          method: 'POST',
          body: JSON.stringify({
            api_key: this.props.user,
            survey:
              {
                surveyName: survey.surveyName,
                surveyExpiration: survey.surveyExpiration,
                questions: survey.questions,
                groups: [{name: cohort_id, members_ids: membersIds}]
              }
          }),
          headers: {
            'Content-Type': 'application/json'
          }
      }
      this.props.handlePost(url, options)
      this.handleSuccess()
    }
  }

  handleSuccess = () => {
    cogoToast.success('Your survey has been sent', {position: 'bottom-left'})
    this.props.history.push('/dashboard')
  }

  sendToast = (message) => {
    cogoToast.warn(message, {position: 'bottom-left'})
  }

  onDrag = (e, student) => {
    e.preventDefault()
    this.setState({
      draggedStudent: {id: student.id, name: student.name}
    })
  }

  onDragOver = (e) => {
    e.preventDefault()
  }

  onDrop = (e) => {
    e.preventDefault()
    const { group, draggedStudent } = this.state
    this.setState({
      group: [...group, draggedStudent ],
      draggedStudent: {}
    })
    const leftoverStudents = this.props.currentCohort.filter(student => {
      return student.id !== draggedStudent.id
    })
    this.props.setCurrentCohort(leftoverStudents)
  }

  render() {
    const studentsToDisplay = this.props.currentCohort.map(student => {
      return <div
              key={student.id}
              id={student.id}
              className="student-nametag"
              draggable onDrag={(e) => this.onDrag(e, student)}>
              {student.name}
             </div>
    })
    const cohortList = this.props.cohorts.map(cohort => {
      return <option key={cohort.id} value={cohort.name} name="cohort_id" >{cohort.name}</option>
    })
    const groupToDisplay = this.state.group.map(student => {
      return <div key={student.id} id={student.id} className="student-nametag">{student.name}</div>
    })
    return(
      <div className="recipient-controls-wrapper">
        <div className="recipients-form-wrapper">
          <h2 className="recipients-form-title">Select Recipients</h2>
          <div className="student-selector-wrapper">
            <select className="drop-down" onChange={this.handleCohort} >
            <option value="0">Select a cohort</option>
            {cohortList}
            </select>
            <select className="drop-down" onChange={this.handleProgram}>
              <option>Select a program</option>
              <option value="b" >BE</option>
              <option value="f" >FE</option>
              <option value="both" >Both</option>
            </select>
            <button className="recipients-button" onClick={this.handleAssignGroups}>Populate Students</button>
          </div>
        </div>
        <div className="student-groups-wrapper"
             style={{display: this.state.displayTeams}}>
          <div className="groups-wrapper"
                onDrop={e => this.onDrop(e)}
                onDragOver={(e => this.onDragOver(e))}>
            <div className="groups-title">Drag Names Here to Form Groups
            </div>
            <div className="groups">
              {groupToDisplay}
            </div>
          </div>
          <div className="students-display">{studentsToDisplay}
          </div>
        </div>
        <button className="recipients-button"
                onClick={this.postSurvey}
                style={{display: this.state.displayTeams}}>Send Survey
        </button>
      </div>
    )
  }
}

RecipientForm.propTypes = {
  cohorts: PropTypes.array,
  survey: PropTypes.object
}

export const mapStateToProps = (state) => ({
  survey: state.survey,
  cohorts: state.cohorts,
  currentCohort: state.currentCohort,
  user: state.user
})

export const mapDispatchToProps = (dispatch) => ({
  handlePost: (url, options) => dispatch(handlePost(url, options)),
  setCurrentCohort: (cohort) => dispatch(setCurrentCohort(cohort))
})

export default connect(mapStateToProps, mapDispatchToProps)(RecipientForm)
