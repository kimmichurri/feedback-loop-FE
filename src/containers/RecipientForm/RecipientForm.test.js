import React from 'react'
import { shallow } from 'enzyme'
import { RecipientForm, mapStateToProps, mapDispatchToProps } from './RecipientForm'
import { setCurrentCohort, setInstructorSurveys, setSurveyTeams } from '../../actions'
jest.mock('../../thunks/handlePost')
jest.mock('../../thunks/handleGet')

describe('RecipientForm', () => {
  let mockState
  let wrapper
  let mockEvent
  let mockHistory = { push: jest.fn() }

  beforeEach(() => {
    mockState = {
      survey : {name: 'name', question: 'question'},
      cohorts: [
        { id: 1 ,
        name: "1811" },
        { id: 2,
        name: "1901" }
        ],
      currentCohort: [
        {id: 1, name:'kim'}, {id: 2, name: 'taylor'}, {id: 3, name:'april'}, {id: 4, name:'peter'}
      ],
      handlePost: jest.fn(),
      handleGet: jest.fn(),
      setCurrentCohort: jest.fn(),
      setInstructorSurveys: jest.fn(),
      setSurveyTeams: jest.fn()
    }

    wrapper = shallow(
      <RecipientForm
        survey={mockState.survey}
        cohorts={mockState.cohorts}
        currentCohort={mockState.currentCohort}
        handleGet={mockState.handleGet}
        handlePost={mockState.handlePost}
        setCurrentCohort={mockState.setCurrentCohort}
        history={mockHistory}
        setInstructorSurveys={mockState.setInstructorSurveys}
        setSurveyTeams={mockState.setSurveyTeams}
      />
    )
  })

  it.skip('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have proper default state', () => {
    expect(wrapper.state('cohort_id')).toEqual(0)
    expect(wrapper.state('program')).toBe('both')
    expect(wrapper.state('draggedStudent')).toEqual({})
    expect(wrapper.state('group')).toEqual([])
    expect(wrapper.state('teams')).toHaveLength(1)
  })

  it('should set state with the program', () => {
    mockEvent = { target: { value: 'f'} }
    wrapper.instance().handleProgram(mockEvent)
    expect(wrapper.state('program')).toEqual('f')
  })

  it('should set state with the cohort id', () => {
    mockEvent = { target: { value: "19"} }
    wrapper.instance().handleCohort(mockEvent)
    expect(wrapper.state('cohort_id')).toEqual('19')
  })

  it('should dispatch handleGet when handleAssignGroups is called', async () => {
    wrapper.instance().handleAssignGroups()
    expect(wrapper.instance().props.handleGet).toHaveBeenCalled()
  })

  it('should dispatch setCurrentCohort when handleAssignGroups is called', async () => {
    wrapper.instance().handleAssignGroups()
    const expected = await wrapper.instance().props.handleGet()
    expect(wrapper.instance().props.setCurrentCohort).toHaveBeenCalledWith(expected)
  })

  it('should dispatch setSurveyTeams when validateGroups is called', async () => {
    const formattedGroups = [{ name: 'team name', members_ids: [99, 4]}];
    wrapper.instance().validateGroups(formattedGroups)
    expect(wrapper.instance().props.setSurveyTeams).toHaveBeenCalled()
  })

  it('should NOT dispatch setSurveyTeams when formatGroups is called with only one team member', async () => {
    const formattedGroups = [{ name: 'team name', members_ids: [99]}];
    const mockState = {
      teams: [{ id: 17, name: 'team one', members: [ {id: 99, name: 'April' }]}]
    }
    wrapper.setState(mockState)
    wrapper.instance().formatGroups()
    expect(wrapper.instance().props.setSurveyTeams).not.toHaveBeenCalled()
  })

  it.skip('should push to dashboard when handleSuccess is called', async () => {
    wrapper.instance().handleSuccess()
    await wrapper.instance().props.handleGet()
    expect(wrapper.instance().props.setInstructorSurveys).toHaveBeenCalled()
    expect(wrapper.instance().props.history.push).toHaveBeenCalled()
  })

  it('should set state with a student when onDrag is invoked', () => {
    const mockEvent = { preventDefault: jest.fn() }
    const mockStudent = { id: 99, name: 'April' }
    wrapper.instance().onDrag(mockEvent, mockStudent)
    expect(wrapper.state('draggedStudent')).toEqual({
      id: 99,
      name: "April"
    })
  })

  it('should prevent default when student is dragged when onDragOver is invoked', () => {
    const mockEvent = { preventDefault: jest.fn() }
    wrapper.instance().onDragOver(mockEvent)
    expect(mockEvent.preventDefault).toHaveBeenCalled()
  })

  it('should update teams and call setCurrentCohort with leftover students when onDrop is invoked', () => {
    const mockEvent = { preventDefault: jest.fn(), target: {id: 17} }
    const mockState = {
      teams: [{ id: 17, name: 'team one', members: []}],
      draggedStudent: { id: 99, name: 'April'}
    }
    const expected = {
      teams: [{ id: 17, name: 'team one', members: [ {id: 99, name: 'April' }]}],
      draggedStudent: {}
    }
    wrapper.setState({
      teams: mockState.teams,
      draggedStudent: mockState.draggedStudent
    })
    wrapper.instance().onDrop(mockEvent)
    expect(wrapper.state('teams')).toEqual(expected.teams)
    expect(wrapper.instance().props.setCurrentCohort).toHaveBeenCalled()
  })

  it('should add a team to state when addTeam is called', () => {
    wrapper.setState({
      teams: [{id: 7, name: 'team one', members: []}]
    })
    wrapper.instance().addTeam()
    expect(wrapper.state('teams')).toHaveLength(2)
  })

  it('should update a team in state when handleTeamName is called', () => {
    const mockState = {
      teams: [{ id: 17, name: 'team one', members: [ {id: 99, name: 'April' }]}]
    }
    const mockTeam = { id: 17, name: 'new name' }
    const expected = [{ id: 17, name: 'new name', members: [ {id: 99, name: 'April' }]}]
    wrapper.setState(mockState)
    wrapper.instance().handleTeamName(mockTeam.id, mockTeam.name)
    expect(wrapper.state('teams')).toEqual(expected)
  })

  it('should call checkTeamNames when checkSurvey is called', () => {
    const checkTeamNamesSpy = jest.spyOn(wrapper.instance(), 'checkTeamNames')
    wrapper.instance().checkSurvey()
    expect(checkTeamNamesSpy).toHaveBeenCalled()
  })

  it('should call formatGroups when checkTeamNames is called', () => {
    const formatGroupsSpy = jest.spyOn(wrapper.instance(), 'formatGroups')
    const mockState = {
      teams: [{ id: 17, name: 'team one', members: [ {id: 99, name: 'April' }]}]
    }
    wrapper.setState({
      teams: mockState.teams
    })
    wrapper.instance().checkTeamNames()
    expect(formatGroupsSpy).toHaveBeenCalled()
  })

  it('should not call validateGroups if a team is missing a name', async () => {
    const validateGroupsSpy = jest.spyOn(wrapper.instance(), 'validateGroups')
    const mockState = {
      teams: [{ id: 17, name: '', members: [ {id: 99, name: 'April' }]}]
    }
    wrapper.setState({
      teams: mockState.teams
    })
    await wrapper.instance().checkTeamNames()
    await expect(validateGroupsSpy).not.toHaveBeenCalled()
  })

  it('formatGroups should call validateGroups with formattedGroups', async () => {
    const validateGroupsSpy = jest.spyOn(wrapper.instance(), 'validateGroups')
    const mockState = {
      teams: [{ id: 17, name: 'team name', members: [ {id: 99, name: 'April' }, {id: 4, name: 'Taylor'}]}]
    }
    const formattedGroups = [{ name: 'team name', members_ids: [99, 4]}]
    await wrapper.setState({
      teams: mockState.teams
    })
    await wrapper.instance().formatGroups()
    await expect(validateGroupsSpy).toHaveBeenCalledWith(formattedGroups)
  })

  it('should not call setSurveyTeams if a team does not have at least two members', () => {
    const setSurveyTeamsSpy = jest.spyOn(wrapper.instance().props, 'setSurveyTeams')
    const mockState = {
      teams: [{ id: 17, name: 'team name', members: [ {id: 99, name: 'April' }]}]
    }
    const formattedGroups = [{ name: 'team name', members_ids: [99]}]
    wrapper.setState({
      teams: mockState.teams
    })
    wrapper.instance().formatGroups()
    expect(setSurveyTeamsSpy).not.toHaveBeenCalledWith(formattedGroups)
  })



  describe('mapStateToProps', () => {
    it('should return the expected state as props', () => {
      mockState = {
        cohorts: [
          { id: 1 ,
          name: "1811" },
          { id: 2,
          name: "1901" }
          ],
        fakeState: "fakeState"
      }
      const expected = {
        cohorts: [
          { id: 1 ,
          name: "1811" },
          { id: 2,
          name: "1901" }
          ]
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should return handlePost to dispatch', () => {
      const mockUrl = 'www.post.com'
        const mockOptions = {
          method: 'POST',
          body: { name: "name",
                  email: "email@gmail.com"
                },
          headers: {
            'Content-Type': 'application/json'
          }
        }
        const mockDispatch = jest.fn()
        const handlePost = jest.fn()
        const actionToDispatch = handlePost(mockUrl, mockOptions)
        const mappedProps = mapDispatchToProps(mockDispatch)

        mappedProps.handlePost(mockUrl, mockOptions)

        expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should return handleGet to dispatch', () => {
      const mockUrl = 'www.post.com'

      const mockDispatch = jest.fn()
      const handleGet = jest.fn()
      const actionToDispatch = handleGet(mockUrl)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.handleGet(mockUrl)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should return setCurrentCohort to dispatch', () => {
      const mockCohort = [
        { id: 1, name: 'Kim'},
        { id: 2, name: 'Taylor'}
      ]
      const mockDispatch = jest.fn()
      const actionToDispatch = setCurrentCohort(mockCohort)
      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.setCurrentCohort(mockCohort)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })

})
