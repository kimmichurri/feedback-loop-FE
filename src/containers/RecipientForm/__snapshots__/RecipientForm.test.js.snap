// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipientForm should match the snapshot 1`] = `
<div
  className="recipient-controls-wrapper"
>
  <div
    className="recipients-form-wrapper"
  >
    <h2
      className="recipients-form-title"
    >
      Select Recipients
    </h2>
    <h3
      className="student-selector"
    >
      Program
    </h3>
    <select
      className="drop-down"
      onChange={[Function]}
    >
      <option
        value="b"
      >
        BE
      </option>
      <option
        value="f"
      >
        FE
      </option>
    </select>
    <h3
      className="student-selector"
    >
      Cohort
    </h3>
    <select
      className="drop-down"
      onChange={[Function]}
    >
      <option
        value="0"
      >
        select a cohort
      </option>
      <option
        key="1"
        name="cohort_id"
        value="1811"
      >
        1811
      </option>
      <option
        key="2"
        name="cohort_id"
        value="1901"
      >
        1901
      </option>
    </select>
    <button
      className="recipients-button"
      onClick={[Function]}
    >
      Assign Groups
    </button>
  </div>
  <div
    className="students-display"
  >
    <div
      className="student-nametag"
      draggable={true}
      id={1}
      key="1"
      onDrag={[Function]}
    >
      kim
    </div>
    <div
      className="student-nametag"
      draggable={true}
      id={2}
      key="2"
      onDrag={[Function]}
    >
      taylor
    </div>
    <div
      className="student-nametag"
      draggable={true}
      id={3}
      key="3"
      onDrag={[Function]}
    >
      april
    </div>
    <div
      className="student-nametag"
      draggable={true}
      id={4}
      key="4"
      onDrag={[Function]}
    >
      peter
    </div>
  </div>
  <div
    className="groups-wrapper"
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <div>
      Drag Names Here to Make a Group
    </div>
  </div>
  <button
    className="recipients-button"
    disabled={true}
    onClick={[Function]}
  >
    Send
  </button>
</div>
`;
